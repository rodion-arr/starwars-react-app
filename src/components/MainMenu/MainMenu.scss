.burger-menu {
  box-sizing: border-box;
  width: 60px;
  padding: 10px;
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  border-right: solid 1px var(--border-color-light);
  cursor: pointer;

  &__bar {
    border: solid 2px var(--border-color-light);
  }
}

.sidebar {
  position: fixed;
  top: 0;
  left: 0;
  display: none;
  opacity: 0;
  height: 100vh;
  transition: opacity .5s ease-in-out;

  .side-menu-opened & {
    display: flex;
    opacity: 1;
  }

  &__menu {
    overflow-y: scroll;
    background-color: #000000;
    width: var(--sidemenu-width);
    height: 100vh;
  }

  &__overlay {
    width: 100vw;
    height: 100vh;
    background-color: #000000;
    opacity: 0.5;
    cursor: pointer;
  }

  .sidebar-menu {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
  }

  .side-menu-items {
    padding: 50px 0 0 25px;

    &__item {
      text-transform: uppercase;
      font-weight: bold;
      font-size: 24px;
      margin-top: 10px;
    }

    &__link {
      color: var(--main-color);
      text-decoration: none;

      &--active {
        text-decoration: underline;
      }
    }
  }
}
