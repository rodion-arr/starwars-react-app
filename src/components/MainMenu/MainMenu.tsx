import './MainMenu.scss';
import React, {useEffect, useState} from "react";
import ReactDOM from 'react-dom';
import {PortalsEnum} from "../../enums/portals.enum";
import {SocialLinks} from "../SocialLinks/SocialLinks";
import {NavLink} from "react-router-dom";
import {URI} from "../../routes/uri";

export const MainMenu = (): JSX.Element => {
  const [sideMenuTarget, setSideMenuTarget] = useState<HTMLElement | null>(null);
  useEffect(() => {
    setSideMenuTarget(document.getElementById(PortalsEnum.SIDE_MENU));

    return () => {
      closeMenu();
    }
  }, []);

  const openMenu = () => {
    document.body.classList.add('side-menu-opened');
  }

  const closeMenu = () => {
    document.body.classList.remove('side-menu-opened');
  }

  return (
    <>
      <div className="header__menu burger-menu" onClick={openMenu}>
        <div className="burger-menu__bar"/>
        <div className="burger-menu__bar"/>
        <div className="burger-menu__bar"/>
      </div>
      {sideMenuTarget && ReactDOM.createPortal(
        (
          <div className="sidebar">
            <div className="sidebar__menu sidebar-menu">
              <div className="sidebar-menu__items side-menu-items">
                <div className="side-menu-items__item"><NavLink to={URI.home} className="side-menu-items__link" activeClassName="side-menu-items__link--active" exact>Home</NavLink></div>
                <div className="side-menu-items__item"><NavLink to={URI.people.base} className="side-menu-items__link" activeClassName="side-menu-items__link--active">People</NavLink></div>
                <div className="side-menu-items__item"><NavLink to={URI.planets.base} className="side-menu-items__link" activeClassName="side-menu-items__link--active">Planets</NavLink></div>
                <div className="side-menu-items__item"><NavLink to={URI.species.base} className="side-menu-items__link" activeClassName="side-menu-items__link--active">Species</NavLink></div>
                <div className="side-menu-items__item"><NavLink to={URI.vehicles} className="side-menu-items__link" activeClassName="side-menu-items__link--active">Vehicles</NavLink></div>
                <div className="side-menu-items__item"><NavLink to={URI.films.base} className="side-menu-items__link" activeClassName="side-menu-items__link--active">Films</NavLink></div>
                <div className="side-menu-items__item"><NavLink to={URI.starships} className="side-menu-items__link" activeClassName="side-menu-items__link--active">Starships</NavLink></div>
              </div>
              <div className="sidebar-menu__social">
                <SocialLinks />
              </div>
            </div>
            <div className="sidebar__overlay" onClick={closeMenu}/>
          </div>
        ),
        sideMenuTarget
      )}
    </>
  );
};
