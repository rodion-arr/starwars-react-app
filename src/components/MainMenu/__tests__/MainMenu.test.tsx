import {render, fireEvent} from "@testing-library/react";
import {MainMenu} from "../MainMenu";
import { BrowserRouter as Router } from "react-router-dom"

describe('MainMenu', () => {
  it('should render correctly', function () {
    const {container} = render(<MainMenu />);

    expect(container).toMatchSnapshot();
  });

  it('should open on burger menu click', function () {
    const {container} = render(<MainMenu />);
    const burgerMenu = container.getElementsByClassName('burger-menu')[0];

    fireEvent.click(burgerMenu);

    expect(document.body).toHaveClass('side-menu-opened');
  });

  it('should close menu page area click', function () {
    const portalRoot = document.createElement('div');
    portalRoot.setAttribute('id', 'side-menu-portal');
    document.body.appendChild(portalRoot)

    const {container, unmount} = render(<Router><MainMenu /></Router>);

    const burgerMenu = container.getElementsByClassName('burger-menu')[0];
    const overlay = document.getElementsByClassName('sidebar__overlay')[0];

    fireEvent.click(burgerMenu);
    fireEvent.click(overlay);

    expect(document.body).not.toHaveClass('side-menu-opened');
    unmount();
  });
});
