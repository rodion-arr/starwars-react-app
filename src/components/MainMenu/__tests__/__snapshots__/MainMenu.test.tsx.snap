// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainMenu should render correctly 1`] = `
<div>
  <div
    class="header__menu burger-menu"
  >
    <div
      class="burger-menu__bar"
    />
    <div
      class="burger-menu__bar"
    />
    <div
      class="burger-menu__bar"
    />
  </div>
</div>
`;
